### PR TITLE
Respect MongoDB max message size

### LINKED ISSUE #21286: MongoDB use maxMessageSize value from hello handshake
### What
MongoDB max message size is a dynamic value set on the server side. Server propagates this value to clients during MongoDB hello handshake:  [maxMessageSize](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/connection.go#L452)

Teleport MongoDB Engine should respect the `maxMessageSize` value received from the mongoDB server and instead using default value for maxMessageSize validate the max message size based on the actual server maxMessageSize value.

### PR DESCRIPTION
**Pull request by gabrielcorado**
Closes #21286.

Read the `maxMessageBytes` value from the MongoDB server (through handshake response) and use it as a maximum size when reading protocol messages.

We're "intercepting" handshake response because Teleport uses a no-op handshake when connecting to the server. This means we cannot access the handshake information using `serverConn.Description().MaxMessageSize`.

### CODE REVIEW

File: lib/srv/db/mongodb/engine.go [hunk 1]
```diff
@@ -159,6 +169,32 @@ func (e *Engine) handleClientMessage(ctx context.Context, sessionCtx *common.Ses
...
 }
 
+// processHandshakeResponse process handshake message and set engine values.
+func (e *Engine) processHandshakeResponse(ctx context.Context, respMessage protocol.Message) {
+	var rawMessage bson.Raw
+	switch resp := respMessage.(type) {
+	// OP_REPLY is used on legacy handshake messages (deprecated on MongoDB 5.0)
+	case *protocol.MessageOpReply:
+		// Handshake messages are always the first document on a reply.
+		rawMessage = bson.Raw(resp.Documents[0])
```
Comment (GavinFrazar) [inline lib/srv/db/mongodb/engine.go:179]: can we check the length of documents here, to avoid a panic if it's 0?

File: lib/srv/db/mongodb/protocol/message.go [hunk 2]
```diff
@@ -101,8 +102,8 @@ func readHeaderAndPayload(reader io.Reader) (*MessageHeader, []byte, error) {
 		return nil, nil, trace.BadParameter("invalid header size %v", header)
 	}
 
-	payloadLength := int64(length - headerSizeBytes)
-	if payloadLength >= defaultMaxMessageSizeBytes {
+	payloadLength := uint32(length - headerSizeBytes)
+	if payloadLength >= maxMessageSize {
```
Comment (GavinFrazar) [inline lib/srv/db/mongodb/protocol/message.go:106]: why the change from 64bit to 32?

Comment (gabrielcorado) [reply lib/srv/db/mongodb/protocol/message.go:106]: I couldn't find documentation mentioning the field type, so I'm assuming [what the driver uses](https://github.com/mongodb/mongo-go-driver/blob/59f7519e8b7c4e549a39ce769b8536f2bdfd0446/mongo/description/server.go#L51).

**Review by GavinFrazar — APPROVED** [review]

**Review by smallinsky — APPROVED** [review]

**Comment by gabrielcorado** [comment]
I will backport this to versions with `// TODO(jent): get the max limit from Mongo handshake`: `v11`, `v12`, `v13`, and `v14`.
